apps/pearl-api/pages/web3auth/swap-owner-session.tsx

```diff
@@ -0,0 +1,371 @@
+          {
+            event_id: Events.WEB3AUTH_SWAP_OWNER_MODAL_CLOSED,
+            success: false,
+            error: 'User closed Web3Auth modal without connecting',
+          },
+          '*',
+        );
+      }
+    };
+
+    web3Auth.on('MODAL_VISIBILITY', handleModalClose);
+    return () => {
+      web3Auth.off('MODAL_VISIBILITY', handleModalClose);
+    };
+  }, [web3Auth, isConnected, result, targetWindow]);
+
+  // Notify if user closes the entire window/iframe before transaction completes
+  useEffect(() => {
+    if (typeof window === 'undefined') return;
+
+    const handleBeforeUnload = () => {
+      // Only send message if transaction hasn't completed yet
+      if (result) return;
+      if (!targetWindow) return;
+      targetWindow.postMessage(
+        {
+          event_id: Events.WEB3AUTH_SWAP_OWNER_MODAL_CLOSED,
+          success: false,
+          error: 'User closed window before transaction completed',
+        },
+        '*',
+      );
+    };
+
+    window.addEventListener('beforeunload', handleBeforeUnload);
+    return () => {
+      window.removeEventListener('beforeunload', handleBeforeUnload);
+    };
+  }, [result, targetWindow]);
+
+  const chainName = useMemo(() => {
+    if (!chainId) return 'Unknown Chain';
+    return `${EvmChainName[chainId as unknown as EvmChainId] || chainId}`;
+  }, [chainId]);
+
+  if (!isInitialized) return <Loading />;
+
+  if (initError) {
+    return (
+      <Alert
+        message="Error Initializing Web3Auth"
+        description={
+          <Text>
+            {(initError instanceof Error ? initError : new Error('Unknown error')).message}
+          </Text>
+        }
+        type="error"
+        icon={<CloseCircleOutlined />}
+        showIcon
+        style={{ margin: 16 }}
+      />
+    );
+  }
+
+  return (
+    <Flex vertical gap={16} style={{ padding: 16 }}>
+      <Title level={3} style={{ margin: 0 }}>
+        {`Approve Transaction - ${chainName} `}
+      </Title>
+
+      <Card>
+        <Space direction="vertical" size="middle" style={{ width: '100%' }}>
+          <div>
+            <Text strong>Status: </Text>
+            {!result && <LoadingOutlined style={{ margin: '0 8px' }} />}
+            <Text>{status}</Text>
+          </div>
+
+          {safeAddress && (
+            <Flex vertical gap={2}>
+              <Text type="secondary">Safe Address:</Text>
+              <Text code copyable>
+                {safeAddress}
+              </Text>
+            </Flex>
+          )}
+        </Space>
+      </Card>
+
+      {result && result.success && (
+        <Alert
+          message="Transaction Successful!"
+          description={
+            <Space direction="vertical" size="small" style={{ width: '100%' }}>
+              <Flex vertical gap={2}>
+                <Text type="secondary">Transaction Hash:</Text>
+                <Text code copyable>
+                  {result.txHash}
+                </Text>
```

Comment:
could be anchor here, but globally I think we are mixing to approaches here - displaying all statuses/errors right in the iframe, and communicating the same to the parent via messages. maybe we can keep showing all here and just communicate to the window booleans "success: true"/"error: true"?